client/tso_client.go

```diff
@@ -55,5 +69,233 @@ func (c *client) GetTSWithinKeyspaceAsync(ctx context.Context, keyspaceID uint32
+	// dc-location -> TSO allocator primary URL
+	tsoAllocators sync.Map // Store as map[string]string
+
+	// primarySwitchedCallbacks will be called after the primary swichted
+	primarySwitchedCallbacks []func()
+	// membersChangedCallbacks will be called after there is any membership
+	// change in the primary and followers
+	membersChangedCallbacks []func()
+	// tsoAllocatorLeaderSwitchedCallback will be called when any keyspace group tso
+	// allocator primary is switched.
+	tsoAllocatorLeaderSwitchedCallback []func()
+
+	checkMembershipCh chan struct{}
+
+	wg     *sync.WaitGroup
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	tlsCfg *tlsutil.TLSConfig
+
+	// Client option.
+	option *option
+}
+
+// newTSOBaseClient returns a new baseClient.
+func newTSOBaseClient(ctx context.Context, cancel context.CancelFunc,
+	wg *sync.WaitGroup, urls []string, tlsCfg *tlsutil.TLSConfig, option *option) BaseClient {
+	bc := &tsoBaseClient{
+		checkMembershipCh: make(chan struct{}, 1),
+		ctx:               ctx,
+		cancel:            cancel,
+		wg:                wg,
+		tlsCfg:            tlsCfg,
+		option:            option,
+	}
+	bc.urls.Store(urls)
+	// TODO: fill the missing part for service discovery
+	bc.switchPrimary(urls)
+
+	_, err := bc.GetOrCreateGRPCConn(bc.getPrimaryAddr())
+	if err != nil {
+		return nil
+	}
+
+	return bc
+}
+
+// Init initialize the concrete client underlying
+func (c *tsoBaseClient) Init() error {
+	return nil
+}
+
+// Close all grpc client connnections
+func (c *tsoBaseClient) CloseClientConns() {
+	c.clientConns.Range(func(_, cc interface{}) bool {
+		if err := cc.(*grpc.ClientConn).Close(); err != nil {
+			log.Error("[pd] failed to close gRPC clientConn", errs.ZapError(errs.ErrCloseGRPCConn, err))
+		}
+		return true
+	})
+}
+
+// GetClusterID returns the ID of the cluster
+func (c *tsoBaseClient) GetClusterID(context.Context) uint64 {
+	return 0
+}
+
+// GetURLs returns the URLs of the servers.
+// For testing use. It should only be called when the client is closed.
+func (c *tsoBaseClient) GetURLs() []string {
+	return c.urls.Load().([]string)
+}
+
+// GetTSOAllocators returns {dc-location -> TSO allocator primary URL} connection map
+func (c *tsoBaseClient) GetTSOAllocators() *sync.Map {
+	return &c.tsoAllocators
+}
+
+// GetTSOAllocatorServingAddrByDCLocation returns the tso allocator of the given dcLocation
+func (c *tsoBaseClient) GetTSOAllocatorServingAddrByDCLocation(dcLocation string) (string, bool) {
+	url, exist := c.tsoAllocators.Load(dcLocation)
+	if !exist {
+		return "", false
+	}
+	return url.(string), true
+}
+
+// GetTSOAllocatorClientConnByDCLocation returns the tso allocator grpc client connection
+// of the given dcLocation
+func (c *tsoBaseClient) GetTSOAllocatorClientConnByDCLocation(dcLocation string) (*grpc.ClientConn, string) {
+	url, ok := c.tsoAllocators.Load(dcLocation)
+	if !ok {
+		panic(fmt.Sprintf("the allocator leader in %s should exist", dcLocation))
```

Comment:
how about log.Fatal?

Reply:
This is the old code introduced by previous commit made 2 years ago.